File: travel-agency/user-management/src/main/java/com/co/unitravel/application/exceptions/general/ApiException.java
package com.co.unitravel.application.exceptions.general;

import lombok.Getter;
import org.springframework.context.support.ResourceBundleMessageSource;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

@Getter
public class ApiException extends Exception {

    private final String message;
    private final String code;
    private final List<String> errors;
    private static final ResourceBundleMessageSource messageSource = new ResourceBundleMessageSource();
    private static final Locale LOCALE_DEFAULT = Locale.ENGLISH;

    public ApiException(String message, String code) {
        super(message);
        messageSource.setBasename("messages");
        messageSource.setDefaultEncoding("UTF-8");
        this.message = getMessage(message, null);
        this.code = getMessage(code, null);
        this.errors = new ArrayList<>();
    }

    public void addError(String error, Object[] params) {
        String errorFound = getMessage(error, params);
        this.errors.add(errorFound);
    }

    public String getMessage(String errorRoute, Object[] params) {
        return messageSource.getMessage(errorRoute, params, LOCALE_DEFAULT);
    }
}

File: travel-agency/reservation-management/src/main/kotlin/com/co/unitravel/application/exceptions/general/ApiException.kt
package com.co.unitravel.application.exceptions.general

import lombok.Getter
import org.springframework.context.support.ResourceBundleMessageSource
import java.util.*
import kotlin.collections.ArrayList

@Getter
open class ApiException(
    override var message: String,
    open var code: String,
) : Exception(message) {

    private val localeDefault = Locale.ENGLISH;

    private val messageSource = ResourceBundleMessageSource()

    private val errors: MutableList<String> = ArrayList()

    fun addError(errorRoute: String, params: Array<Any>?) {
        messageSource.setBasename("messages")
        messageSource.setDefaultEncoding("UTF-8")
        processInfo()
        val error = getMessage(errorRoute, params)
        errors.add(error)
    }

    fun errors(): MutableList<String> {
        return errors
    }

    private fun processInfo(){
        message = getMessage(message, null)
        code = getMessage(code, null)
    }

    private fun getMessage(route:String, params: Array<Any>?): String{
        return messageSource.getMessage(route, params, localeDefault);
    }
}